### DISCUSSION TITLE
Possibility about implementing ML models with webassembly on IoT device

### DISCUSSION THREAD

**Discussion by TAOFOR4** [bytecodealliance/wasm-micro-runtime#3221] [Q&A]
Hi, I have recently been doing research for implementing ML models with webassembly on IoT devices. I just noticed wasi-nn, but it seems only available for Intel processors now. Do you know about IoT devices? Can I use wamr to run ML models on IoT devices?

**Comment by abrown** [comment]
See my response to your same question here: https://github.com/WebAssembly/wasi-nn/discussions/68. No, wasi-nn is not available only on Intel processors.

**Reply by TAOFOR4** [reply]
Thank you for getting back to me. But is WAMR available to run wasi-nn on IoT devices like raspberry pi or so now? If so, could you provide some resource for me? I would like to do some testing for this part, but I failed to find any documentation.

**Comment by wenyongh** [comment]
@TAOFOR4 we haven't tested WAMR wasi-nn on raspberry pi yet, please refer to:
https://github.com/bytecodealliance/wasm-micro-runtime/tree/main/core/iwasm/libraries/wasi-nn

@tonibofarull maybe you have experience about running WAMR wasi-nn in the IoT device, is there anything you can share? Thanks.

**Reply by tonibofarull** [reply]
Yes, I have tested wasi-nn on Rasp and it works fine. Any concrete error that you are getting? @TAOFOR4

**Reply by TAOFOR4** [reply]
> Yes, I have tested wasi-nn on Rasp and it works fine. Any concrete error that you are getting? @TAOFOR4 

Hi! Thx for telling me that. Actually I haven't tested wasi-nn on raspberry pi yet. I am just going to buy some devices for testing so come to ask for possibilities. Also, do you know if it is possible to run wasi-nn on tiny devices like raspberry pico ,ESP32 C6 or something similar? Thank you!

**Reply by tonibofarull** [reply]
Should run fine in Rasp Pico. However, I'd recommend Coral USB or Intel Neural Compute Stick (or any other TPU) if you want real time performance. In my case, I have tested Coral.
Haven't tested ESP32 C6, or any other RISC CPU, yet.

**Reply by TAOFOR4** [reply]
> Should run fine in Rasp Pico. However, I'd recommend Coral USB or Intel Neural Compute Stick (or any other TPU) if you want real time performance. In my case, I have tested Coral.
> 
> Haven't tested ESP32 C6, or any other RISC CPU, yet.

Got it! Thank you! I will try that. And also appreciate the answers for others :)

**Reply by TAOFOR4** [reply]
> Should run fine in Rasp Pico. However, I'd recommend Coral USB or Intel Neural Compute Stick (or any other TPU) if you want real time performance. In my case, I have tested Coral.
> 
> Haven't tested ESP32 C6, or any other RISC CPU, yet.

Hi! Also, I am wondering if wamr supports pico. Based on my research, I only saw wasm3 support s pico. If only wasm3 works for pico, does it mean I can only use wasm3 with wasi-nn on pico? Thx
